transport_analysis/viscosity.py

```diff
@@ -0,0 +1,191 @@
+viscosity via the Einstein-Helfand method.
+
+"""
+from typing import TYPE_CHECKING
+
+from MDAnalysis.analysis.base import AnalysisBase
+from MDAnalysis.core.groups import UpdatingAtomGroup
+from MDAnalysis.exceptions import NoDataError
+import numpy as np
+
+if TYPE_CHECKING:
+    from MDAnalysis.core.universe import AtomGroup
+
+
+class ViscosityHelfand(AnalysisBase):
+    """
+    Class to calculate viscosity using the Einstein-Helfand approach.
+
+    Parameters
+    ----------
+    atomgroup : AtomGroup
+        An MDAnalysis :class:`~MDAnalysis.core.groups.AtomGroup`.
+        Note that :class:`UpdatingAtomGroup` instances are not accepted.
+    vol_avg : float
+        Average volume over the course of the simulation.
+    temp_avg : float (optional, default 300)
+        Average temperature over the course of the simulation.
+    dim_type : {'xyz', 'xy', 'yz', 'xz', 'x', 'y', 'z'}
+        Desired dimensions to be included in the viscosity calculation.
+        Defaults to 'xyz'.
+
+    Attributes
+    ----------
+    atomgroup : :class:`~MDAnalysis.core.groups.AtomGroup`
+        The atoms to which this analysis is applied
+    dim_fac : int
+        Dimensionality :math:`d` of the viscosity computation.
+    results.timeseries : :class:`numpy.ndarray`
+        The averaged viscosity over all the particles with respect to lag-time.
+        Obtained after calling :meth:`ViscosityHelfand.run`
+    results.visc_by_particle : :class:`numpy.ndarray`
+        The viscosity of each individual particle with respect to lag-time.
+    start : Optional[int]
+        The first frame of the trajectory used to compute the analysis.
+    stop : Optional[int]
+        The frame to stop at for the analysis.
+    step : Optional[int]
+        Number of frames to skip between each analyzed frame.
+    n_frames : int
+        Number of frames analysed in the trajectory.
+    n_particles : int
+        Number of particles viscosity was calculated over.
+    """
+
+    def __init__(
+        self,
+        atomgroup: "AtomGroup",
+        vol_avg: float,
+        temp_avg: float = 300.0,
+        dim_type: str = "xyz",
+        **kwargs,
+    ) -> None:
+        # the below line must be kept to initialize the AnalysisBase class!
+        super().__init__(atomgroup.universe.trajectory, **kwargs)
+        # after this you will be able to access `self.results`
+        # `self.results` is a dictionary-like object
+        # that can should used to store and retrieve results
+        # See more at the MDAnalysis documentation:
+        # https://docs.mdanalysis.org/stable/documentation_pages/analysis/base.html?highlight=results#MDAnalysis.analysis.base.Results
+
+        if isinstance(atomgroup, UpdatingAtomGroup):
+            raise TypeError(
+                "UpdatingAtomGroups are not valid for viscosity computation"
+            )
+
+        # args
+        self.dim_type = dim_type
+        self._parse_dim_type()
```

Comment:
I think this should be a static method.

```python
self.dim_type = dim_type
self._parse_dim_type()
```

is much less clear than

```python
self.dim_type, self.dim_fac = self._parse_dim_type(dim_type)
```

The latter makes it clearer what _parse_dim_type is doing. In the former, you don't even know that a self.dim_fac variable has been created.